### ISSUE TITLE
Replace request with node-fetch

### ISSUE THREAD

**Issue by minas90** [codemanki/cloudscraper#326]
Hi there,
I think it's a good idea to replace `request`, as it's [deprecated](https://github.com/request/request/issues/3142) now.

`node-fetch` seems to be the best future proof option at the moment.

**Comment by kevinvanrijn** [comment]
A rewrite of cloudscraper might take a while seeing how little activity this repo has had, so another short term solution would be to recommend a maintained fork of request.

https://github.com/postmanlabs/postman-request is one example I could find from a quick glace at request's Network tab, which seems to be the most well maintained fork.

---

Agreed on node-fetch as a replacement fwiw.

**Comment by RedNinjaX** [comment]
At this rate we may need a cloudscraper replacement as well

**Comment by codemanki** [comment]
@kevinvanrijn @RedNinjaX you are always welcome to make a PR to this repo and help others and me with the rate.

**Comment by kevinvanrijn** [comment]
I can't promise anything for the rewrite side of things (I simply don't have the time for any significant contributions) but I will be looking to move my own request-dependent projects over to a maintained fork.

So if you don't see cloudscraper moving away from request anytime soon then I can make the relevant changes here as well.

I'll first be waiting for the dust to settle though. I want to be sure no other forks pop up and see if the Postman guys are still interested in maintaining their fork.

**Comment by kevinvanrijn** [comment]
I've received confirmation that `postman-request` is alive. https://github.com/request/request/issues/3142#issuecomment-587569671

In addition, one of the maintainers has opened up a [pull request](https://github.com/postmanlabs/postman-request/pulls) on their repo. So it looks like `postman-request` will continue to be maintained for the foreseeable future.

I'll leave the next steps up to you @codemanki , would you be interested in having `cloudscraper` moved over to `postman-request` or is a rewrite coming sooner than expected?

**Comment by codemanki** [comment]
@kevinvanrijn what would be the benefits of replacing request with postman-request as of now? Cloudscraper doesn't use a lot of request features, so ( imho ) there are no critical issues directly related to outdated request library.

**Comment by minas90** [comment]
@codemanki The main reason for deprecating request is the fact that it is not in line with modern js. And at some point it may stop working on a newer version of node after some breaking changes.

My main concern is the fact that in the past every time I ran `npm install` I got exactly 1 warning from dozen or so deps I use and it came from `cloudscraper`. I use ` Node.js v10 or later`, but I get nagged to install `Brotli ` and even for older versions `brotli` is optional I think, right?

And a week ago I started to get 2 warnings, `brotli` and `request` and both come from `cloudscraper` and it's kind of annoying tbh. My automated deployment script does `npm install` and I deploy dozens of times per day. Every time these 2 warnings grab my attention. Of course I can simply tweak the script to hide warnings, but what if there is something else more critical?

**Comment by kevinvanrijn** [comment]
@codemanki It would mainly be to not depend on a deprecated package. It's not just request that isn't getting updates, all of it's dependencies will also never be updated beyond what the current package.json defines. A security issue is bound to happen eventually.

I don't think any of the bugfixes made in postman-request are relevant to cloudscraper but it does fix some edge-cases in request's behavior that may be relevant to cloudscraper users.

If you'd prefer to not make any decisions yet that's also fine with me.

**Comment by dwayne-dev** [comment]
IMHO, request has always been a bulky mess that wasn't maintained properly and could have been considered to be deprecated *long* ago. Many people have tried to fix that and were shot down or silently ignored by the original author. I do not like request at all.

Re outdated dependencies, I expected users of this package to audit their dependencies regardless of whether request is deprecated. There are just too many well known problems with how NPM works for developers not to be doing so. For example, look at some of the past security incidents.

The issues being discussed here are minor to me as there are workarounds i.e. https://classic.yarnpkg.com/en/docs/selective-version-resolutions/ and lockfiles. Concerning breakage on newer versions of Node, that can and does/will happen anyway... If that's really an issue, use CI to automatically test newer versions of node and respond accordingly. It's not an immediate issue for the majority of users since request works just fine on LTS versions of node.

Re annoyances, if support for Node.js 8 is officially dropped, there is no longer a need to alert users to the lack of native support for Brotli or the workaround for that. Node.js 8 is past EOL so I would expect a PR to fix the annoyance to be very welcome.

In summary, there's no need to rush this at all but user's of this project should strongly consider sending PRs to help it's maintainers.

I'm not casting a vote for a particular library right now but +1 for dropping request entirely.

**Comment by dwayne-dev** [comment]
I'm also against switching out request for a similar allegedly maintained version. Cloudscraper has support for overriding the underlying request library. If anything, I'd expect support for that to be improved/abstracted and the direct dependency on any specific request library to be removed.
